File: data_structures/go/sets.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type CustomSet struct {
	elements map[int]bool // Map to mimic a set
}

// Basic Set Operations

// isEmpty checks if the set is empty.
func (set *CustomSet) isEmpty() bool {
	return len(set.elements) == 0
}

// count returns the number of elements in the set.
func (set *CustomSet) count() int {
	return len(set.elements)
}

// insert inserts an element into the set.
func (set *CustomSet) insert(element int) {
	set.elements[element] = true
}

// remove removes an element from the set.
func (set *CustomSet) remove(element int) {
	delete(set.elements, element)
}

// contains checks if the set contains a specific element.
func (set *CustomSet) contains(element int) bool {
	_, exists := set.elements[element]
	return exists
}

// Set Operations

// union performs the union operation with another set and returns a new set.
func (set *CustomSet) union(otherSet *CustomSet) *CustomSet {
	newSet := &CustomSet{elements: make(map[int]bool)}
	for element := range set.elements {
		newSet.insert(element)
	}
	for element := range otherSet.elements {
		newSet.insert(element)
	}
	return newSet
}

// intersection performs the intersection operation with another set and returns a new set.
func (set *CustomSet) intersection(otherSet *CustomSet) *CustomSet {
	newSet := &CustomSet{elements: make(map[int]bool)}
	for element := range set.elements {
		if otherSet.contains(element) {
			newSet.insert(element)
		}
	}
	return newSet
}

// difference performs the difference operation with another set and returns a new set.
func (set *CustomSet) difference(otherSet *CustomSet) *CustomSet {
	newSet := &CustomSet{elements: make(map[int]bool)}
	for element := range set.elements {
		if !otherSet.contains(element) {
			newSet.insert(element)
		}
	}
	return newSet
}

// isSubsetOf checks if this set is a subset of another set.
func (set *CustomSet) isSubsetOf(otherSet *CustomSet) bool {
	for element := range set.elements {
		if !otherSet.contains(element) {
			return false
		}
	}
	return true
}

// isSupersetOf checks if this set is a superset of another set.
func (set *CustomSet) isSupersetOf(otherSet *CustomSet) bool {
	return otherSet.isSubsetOf(set)
}

// isDisjointWith checks if this set is disjoint with another set.
func (set *CustomSet) isDisjointWith(otherSet *CustomSet) bool {
	for element := range set.elements {
		if otherSet.contains(element) {
			return false
		}
	}
	return true
}

// forEach applies a closure to each element in the set.
func (set *CustomSet) forEach(closure func(int)) {
	for element := range set.elements {
		closure(element)
	}
}

// removeAll removes all elements from the set.
func (set *CustomSet) removeAll() {
	set.elements = make(map[int]bool)
}

// removeAllOccurrences removes all occurrences of an element from the set.
func (set *CustomSet) removeAllOccurrences(element int) {
	delete(set.elements, element)
}

// symmetricDifference returns the set with common elements between this set and another set.
func (set *CustomSet) symmetricDifference(otherSet *CustomSet) *CustomSet {
	diff1 := set.difference(otherSet)
	diff2 := otherSet.difference(set)
	return diff1.union(diff2)
}

// formUnion inserts elements from another set into this set.
func (set *CustomSet) formUnion(otherSet *CustomSet) {
	for element := range otherSet.elements {
		set.insert(element)
	}
}

// formIntersection removes elements of this set that are not in another set.
func (set *CustomSet) formIntersection(otherSet *CustomSet) {
	for element := range set.elements {
		if !otherSet.contains(element) {
			delete(set.elements, element)
		}
	}
}

// subtract removes elements of this set that are in another set.
func (set *CustomSet) subtract(otherSet *CustomSet) {
	for element := range otherSet.elements {
		delete(set.elements, element)
	}
}

// ... Add more set operations as needed ...

// PrintSet prints the elements of the set.
func (set *CustomSet) printSet() {
	var elements []string
	for element := range set.elements {
		elements = append(elements, strconv.Itoa(element))
	}
	fmt.Printf("{%s}\n", strings.Join(elements, ", "))
}

File: data_structures/rust/sets.rs
use std::collections::HashMap;

#[derive(Clone)]
struct CustomSet<T: std::hash::Hash + Eq> {
    elements: HashMap<T, bool>,  // HashMap to mimic a set
}

impl<T: std::hash::Hash + Eq> CustomSet<T> {
    // Basic Set Operations

    /// Check if the set is empty.
    fn is_empty(&self) -> bool {
        self.elements.is_empty()
    }

    /// Get the number of elements in the set.
    fn count(&self) -> usize {
        self.elements.len()
    }

    /// Insert an element into the set.
    fn insert(&mut self, element: T) {
        self.elements.insert(element, true);
    }

    /// Remove an element from the set.
    fn remove(&mut self, element: &T) -> Option<T> {
        if self.elements.remove(element).is_some() {
            Some(element.clone())
        } else {
            None
        }
    }

    /// Check if the set contains a specific element.
    fn contains(&self, element: &T) -> bool {
        self.elements.contains_key(element)
    }

    // Set Operations

    /// Perform the union operation with another set and return a new set.
    fn union(&self, other_set: &CustomSet<T>) -> CustomSet<T> {
        let mut new_set = self.clone();
        for element in other_set.elements.keys() {
            new_set.insert(element.clone());
        }
        new_set
    }

    /// Perform the intersection operation with another set and return a new set.
    fn intersection(&self, other_set: &CustomSet<T>) -> CustomSet<T> {
        let mut new_set = CustomSet { elements: HashMap::new() };
        for element in self.elements.keys() {
            if other_set.contains(element) {
                new_set.insert(element.clone());
            }
        }
        new_set
    }

    /// Perform the difference operation with another set and return a new set.
    fn difference(&self, other_set: &CustomSet<T>) -> CustomSet<T> {
        let mut new_set = self.clone();
        for element in other_set.elements.keys() {
            new_set.remove(element);
        }
        new_set
    }

    /// Check if this set is a subset of another set.
    fn is_subset_of(&self, other_set: &CustomSet<T>) -> bool {
        self.elements.keys().all(|element| other_set.contains(element))
    }

    /// Check if this set is a superset of another set.
    fn is_superset_of(&self, other_set: &CustomSet<T>) -> bool {
        other_set.is_subset_of(self)
    }

    /// Check if this set is disjoint with another set.
    fn is_disjoint_with(&self, other_set: &CustomSet<T>) -> bool {
        self.elements.keys().all(|element| !other_set.contains(element))
    }

    /// Apply a closure to each element in the set.
    fn for_each<F>(&self, closure: F)
    where
        F: Fn(&T),
    {
        for element in self.elements.keys() {
            closure(element);
        }
    }

    /// Remove all elements from the set.
    fn remove_all(&mut self) {
        self.elements.clear();
    }

    /// Remove all occurrences of an element from the set.
    fn remove_all_occurrences(&mut self, element: &T) {
        self.elements.remove(element);
    }

    /// Get the set with common elements between this set and another set.
    fn symmetric_difference(&self, other_set: &CustomSet<T>) -> CustomSet<T> {
        let diff1 = self.difference(other_set);
        let diff2 = other_set.difference(self);
        diff1.union(&diff2)
    }

    /// Insert elements from another set into this set.
    fn form_union(&mut self, other_set: &CustomSet<T>) {
        for element in other_set.elements.keys() {
            self.insert(element.clone());
        }
    }

    /// Remove elements of this set that are not in another set.
    fn form_intersection(&mut self, other_set: &CustomSet<T>) {
        let intersection = self.intersection(other_set);
        self.elements = intersection.elements;
    }

    /// Remove elements of this set that are in another set.
    fn subtract(&mut self, other_set: &CustomSet<T>) {
        for element in other_set.elements.keys() {
            self.remove(element);
        }
    }

    // ... Add more set operations as needed ...
}

// Example usage
fn main() {
    let mut set1 = CustomSet {
        elements: HashMap::new(),
    };
    set1.insert(1);
    set1.insert(2);
    set1.insert(3);

    let mut set2 = CustomSet {
        elements: HashMap::new(),
    };
    set2.insert(2);
    set2.insert(3);
    set2.insert(4);

    let union_set = set1.union(&set2);
    println!("Union set: {:?}", union_set.elements.keys());

    let intersection_set = set1.intersection(&set2);
    println!("Intersection set: {:?}", intersection_set.elements.keys());
}
